File: examples/test_docs_simple.py
#!/usr/bin/env python3
"""Example module for documentation generation.

This module demonstrates various Python features that Depyler
can document, including functions and classes.
"""


def fibonacci(n: int) -> int:
    """Calculate the n-th Fibonacci number.

    This function uses an iterative approach for efficiency.

    Args:
        n: The position in the Fibonacci sequence (0-indexed)

    Returns:
        The n-th Fibonacci number
    """
    if n <= 1:
        return n

    a: int = 0
    b: int = 1
    for i in range(2, n + 1):
        temp: int = a + b
        a = b
        b = temp
    return b


def process_data(items: list[int]) -> dict[str, int]:
    """Process a list of integers and return statistics.

    Args:
        items: List of integers to process

    Returns:
        Dictionary containing statistics
    """
    stats: dict[str, int] = {}
    stats["count"] = len(items)
    stats["sum"] = sum(items)

    return stats


class DataProcessor:
    """A class for processing and analyzing data.

    This class provides methods for various data processing operations.
    """

    def __init__(self, name: str):
        """Initialize a new DataProcessor.

        Args:
            name: Name for this processor instance
        """
        self.data: list[int] = []
        self.name: str = name

    def add_data(self, value: int) -> None:
        """Add a single data point.

        Args:
            value: The integer value to add
        """
        self.data.append(value)

    def get_count(self) -> int:
        """Get the count of data points.

        Returns:
            Number of data points
        """
        return len(self.data)

    @staticmethod
    def create_default() -> "DataProcessor":
        """Create a default processor.

        Returns:
            New DataProcessor instance
        """
        return DataProcessor("default")

File: examples/test_docs_simple.rs
use serde_json;
use std::collections::HashMap;
#[derive(Debug, Clone)]
pub struct DataProcessor {
    pub data: Vec<serde_json::Value>,
    pub name: serde_json::Value,
}
impl DataProcessor {
    pub fn new(name: Option<String>) -> Self {
        Self {
            data: Vec::new(),
            name,
        }
    }
    pub fn add_data(&self, value: i32) {
        self.data.push(value);
    }
    pub fn get_count(&self) -> i32 {
        return self.data.len() as i32;
    }
    pub fn create_default() -> DataProcessor {
        return DataProcessor::new("default".to_string());
    }
}
#[doc = "Calculate the n-th Fibonacci number.\n    \n    This function uses an iterative approach for efficiency.\n    \n    Args:\n        n: The position in the Fibonacci sequence(0-indexed)\n        \n    Returns:\n        The n-th Fibonacci number\n    "]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn fibonacci(n: i32) -> i32 {
    let _cse_temp_0 = n <= 1;
    if _cse_temp_0 {
        return n;
    }
    let mut a = 0;
    let mut b = 1;
    for _i in 2..n + 1 {
        let temp = a + b;
        a = b;
        b = temp;
    }
    b
}
#[doc = "Process a list of integers and return statistics.\n    \n    Args:\n        items: List of integers to process\n        \n    Returns:\n        Dictionary containing statistics\n    "]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn process_data(items: Vec<i32>) -> HashMap<String, i32> {
    let mut stats = {
        let map = HashMap::new();
        map
    };
    let _cse_temp_0 = items.len() as i32;
    stats.insert("count".to_string().to_string(), _cse_temp_0);
    let _cse_temp_1 = items.iter().sum::<i32>();
    stats.insert("sum".to_string().to_string(), _cse_temp_1);
    stats
}
#[cfg(test)]
mod tests {
    use super::*;
    use quickcheck::{quickcheck, TestResult};
    #[test]
    fn test_fibonacci_examples() {
        assert_eq!(fibonacci(0), 0);
        assert_eq!(fibonacci(1), 1);
        assert_eq!(fibonacci(-1), -1);
    }
}
